src/search/landmarks/landmark_status_manager.cc

```diff
@@ -8,164 +8,216 @@ using namespace std;
+    assert(past.size() == num_landmarks);
+    assert(parent_past.size() == num_landmarks);
+    assert(fut.size() == num_landmarks);
+    assert(parent_fut.size() == num_landmarks);
 
+    progress_basic(parent_past, parent_fut, parent_ancestor_state, past,
+                   fut, ancestor_state);
 
-    // Mark landmarks reached right now as "reached" (if they are "leaves").
     for (int id = 0; id < num_landmarks; ++id) {
-        if (!reached.test(id)) {
-            LandmarkNode *node = lm_graph.get_node(id);
-            if (node->get_landmark().is_true_in_state(ancestor_state)) {
-                if (landmark_is_leaf(*node, reached)) {
-                    reached.set(id);
+        if (progress_goals) {
+            progress_goal(id, ancestor_state, fut);
+        }
+        if (progress_greedy_necessary_orderings) {
+            progress_greedy_necessary_ordering(id, ancestor_state, past, fut);
+        }
+        if (progress_reasonable_orderings) {
+            progress_reasonable_ordering(id, past, fut);
+        }
+    }
+}
+
+void LandmarkStatusManager::progress_basic(
+    const BitsetView &parent_past, const BitsetView &parent_fut,
+    const State &parent_ancestor_state, BitsetView &past,
+    BitsetView &fut, const State &ancestor_state) {
+    utils::unused_variable(parent_past);
+    utils::unused_variable(parent_ancestor_state);
+    for (auto &node : lm_graph.get_nodes()) {
+        int id = node->get_id();
+        const Landmark &lm = node->get_landmark();
+        if (parent_fut.test(id)) {
+            /*
+              A future landmark remains future if it is not achieved by the
+              current transition. If it additionally wasn't past in the parent,
+              it remains not past.
+
+              We first test whether the landmark is currently considered past or
+              not future. If neither holds, the code that follows the condition
+              has no effect and we can therefore avoid computing whether the
+              landmark holds in the current state.
+              TODO: Computing whether a landmark holds in a state is expensive.
+               It can happen that we compute this multiple times (considering
+               also the ordering progressions that follow) for the same landmark
+               and state. If we observe a slow-down in the experiments, we could
+               look into this as an opportunity to speed things up.
+            */
+            if ((past.test(id) || !fut.test(id))
+                && !lm.is_true_in_state(ancestor_state)) {
+                fut.set(id);
+                if (!parent_past.test(id)) {
+                    past.reset(id);
                 }
             }
         }
     }
-
-    return true;
 }
 
-void LandmarkStatusManager::update_lm_status(const State &ancestor_state) {
-    const BitsetView reached = get_reached_landmarks(ancestor_state);
+void LandmarkStatusManager::progress_goal(
```

Comment:
It might be more efficient to call this function only for goal landmarks (this would require to store either pointers or indices of goal landmarks in the status manager).

Reply:
Similarly, it might be more efficient to loop only over greedy-necessary orderings and reasonable orderings in the according progression functions. Should I compute lookups for these things when the status manager is initialized and store them? Or do you have an even better idea? Doing something in this direction could certainly speed things up in practice, but it doesn't affect big-O.